### PR TITLE
Support setting response header size limits

### PR DESCRIPTION
**Pull request by jorop**
## What do these changes do?

Allow the user to set the response header limits. The properties max_line_size, max_headers and max_field_size can be set on ClientSession creation. Default values are: `max_line_size=8190` `max_headers=32768` and `max_field_size=8190`. 
Thus the user can set and - when needed - receive bigger response headers. 

## Are there changes in behavior for the user?

No

## Related issue number

#2304

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

File: aiohttp/client.py [hunk 1]
```diff
@@ -351,6 +360,9 @@ async def _request(
         proxy_headers: Optional[LooseHeaders] = None,
         trace_request_ctx: Optional[SimpleNamespace] = None,
         read_bufsize: Optional[int] = None,
+        max_line_size: Optional[int] = 8190,
+        max_headers: Optional[int] = 32768,
+        max_field_size: Optional[int] = 8190,
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]: Shouldn't these default to `None`? This currently means that the session defaults are only used when the user explicitly passes `None`, so they would need to make every call like `client.request(..., max_line_size=None, max_headers=None, max_field_size=None)`, making the session defaults pointless.

**Review by Dreamsorcerer — CHANGES_REQUESTED** [review]
Will need some tests for the new behaviour to have any chance of being merged, which would also have highlighted the below issue.

File: aiohttp/client.py (hunk 1 above)
Comment (jorop) [reply aiohttp/client.py]: Thanks for the review. Of course that's right. I changed it accordingly and added a couple of tests.

File: tests/test_client_functional.py [hunk 2]
```diff
@@ -3041,3 +3041,126 @@ async def handler(request):
...
+    app = web.Application()
+    app.add_routes([web.get("/", handler)])
+
+    client = await aiohttp_client(app)
+
+    async with await client.get("/") as resp:
+        assert resp.headers["Custom"] == "x" * 8190
+
+
+@pytest.mark.xfail
```
Comment (Dreamsorcerer) [inline tests/test_client_functional.py]: `xfail` generally means the test is not working yet and requires something to be fixed.

This should be rewritten to test the exact expected behaviour (e.g. using `with pytest.raises(...)`).

**Review by Dreamsorcerer — APPROVED** [review]

**Comment by manuel2295** [comment]
When will this change be approved? This issue is very important to the use of the library.

**Comment by Dreamsorcerer** [comment]
There is a test failure, which I've not looked at closely yet. It's not one I've seen before though.

**Comment by smonist** [comment]
This is the test that failed: [https://github.com/aio-libs/aiohttp/runs/5072020477?check_suite_focus=true#step:10:3041](https://github.com/aio-libs/aiohttp/runs/5072020477?check_suite_focus=true#step:10:3041)
It looks like the plausible timeout for this test is set too narrow. Can we run the test again and look at the results? Seems not related to the code changes here.

**Comment by dmitry-brazhenko** [comment]
Hello @Dreamsorcerer!

Is it possible to merge this PR ? :) 

Thanks

File: CHANGES/2304.feature [hunk 3]
```diff
@@ -0,0 +1 @@
+Support setting response header parameters max_line_size, max_headers and max_field_size.
```
Comment (Dreamsorcerer) [inline CHANGES/2304.feature]:
```suggestion
Support setting response header parameters max_line_size and max_field_size.
```

File: aiohttp/client.py [hunk 4]
```diff
@@ -186,6 +186,9 @@ class ClientSession:
         "_ws_response_class",
         "_trace_configs",
         "_read_bufsize",
+        "_max_line_size",
+        "_max_headers",
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]:
```suggestion
```

File: aiohttp/client.py [hunk 5]
```diff
@@ -213,6 +216,9 @@ def __init__(
         requote_redirect_url: bool = True,
         trace_configs: Optional[List[TraceConfig]] = None,
         read_bufsize: int = 2**16,
+        max_line_size: int = 8190,
+        max_headers: int = 32768,
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]:
```suggestion
```

File: aiohttp/client.py [hunk 6]
```diff
@@ -261,6 +267,9 @@ def __init__(
         self._trust_env = trust_env
         self._requote_redirect_url = requote_redirect_url
         self._read_bufsize = read_bufsize
+        self._max_line_size = max_line_size
+        self._max_headers = max_headers
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]:
```suggestion
```

File: aiohttp/client.py [hunk 7]
```diff
@@ -342,6 +351,9 @@ async def _request(
         trace_request_ctx: Optional[SimpleNamespace] = None,
         read_bufsize: Optional[int] = None,
         auto_decompress: Optional[bool] = None,
+        max_line_size: Optional[int] = None,
+        max_headers: Optional[int] = None,
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]:
```suggestion
```

File: aiohttp/client.py [hunk 8]
```diff
@@ -404,6 +416,15 @@ async def _request(
         if auto_decompress is None:
             auto_decompress = self._auto_decompress
 
+        if max_line_size is None:
+            max_line_size = self._max_line_size
+
+        if max_headers is None:
+            max_headers = self._max_headers
+
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]:
```suggestion
```

File: aiohttp/client.py [hunk 9]
```diff
@@ -510,6 +531,9 @@ async def _request(
                         read_timeout=real_timeout.sock_read,
                         read_bufsize=read_bufsize,
                         timeout_ceil_threshold=self._connector._timeout_ceil_threshold,
+                        max_line_size=max_line_size,
+                        max_headers=max_headers,
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]:
```suggestion
```

File: aiohttp/client.py [hunk 10]
```diff
@@ -1186,6 +1210,9 @@ def request(
     version: HttpVersion = http.HttpVersion11,
     connector: Optional[BaseConnector] = None,
     read_bufsize: Optional[int] = None,
+    max_line_size: int = 8190,
+    max_headers: int = 32768,
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]:
```suggestion
```

File: aiohttp/client.py [hunk 11]
```diff
@@ -1256,6 +1283,9 @@ def request(
             proxy=proxy,
             proxy_auth=proxy_auth,
             read_bufsize=read_bufsize,
+            max_line_size=max_line_size,
+            max_headers=max_headers,
```
Comment (Dreamsorcerer) [inline aiohttp/client.py]:
```suggestion
```

File: aiohttp/client_proto.py [hunk 12]
```diff
@@ -153,6 +153,9 @@ def set_response_params(
         read_timeout: Optional[float] = None,
         read_bufsize: int = 2**16,
         timeout_ceil_threshold: float = 5,
+        max_line_size: int = 8190,
+        max_headers: int = 32768,
```
Comment (Dreamsorcerer) [inline aiohttp/client_proto.py]:
```suggestion
```

File: aiohttp/client_proto.py [hunk 13]
```diff
@@ -169,6 +172,9 @@ def set_response_params(
             response_with_body=not skip_payload,
             read_until_eof=read_until_eof,
             auto_decompress=auto_decompress,
+            max_line_size=max_line_size,
+            max_headers=max_headers,
```
Comment (Dreamsorcerer) [inline aiohttp/client_proto.py]:
```suggestion
```

File: tests/test_client_functional.py [hunk 14]
```diff
@@ -3065,3 +3065,124 @@ async def handler(request):
...
+            headers.add(f"x-header-{x}", str(x))
+        return web.Response(headers=headers)
+
+    app = web.Application()
+    app.add_routes([web.get("/", handler)])
+
+    client = await aiohttp_client(app)
+
+    async with await client.get("/") as resp:
+        assert len(resp.headers) == 32768
```
Comment (Dreamsorcerer) [inline tests/test_client_functional.py]:
```suggestion
```

**Comment by Dreamsorcerer** [comment]
If someone could handle the backport to 3.9 as above, that'd be great.

**Comment by rohansahai** [comment]
Awesome to see this change coming 😃 . Is there any ETA of the 3.9 release?

**Comment by Dreamsorcerer** [comment]
Not yet, but still waiting on the backport to 3.9, if you'd like to help out:
https://github.com/aio-libs/aiohttp/pull/6559#issuecomment-1522189057

**Pull request merged**